frontend/src/stores/useProjectStore.ts

```diff
@@ -226,6 +244,87 @@ const useProjectStore = create<ProjectStore>()(persist((set: SetState<ProjectSto
     project.states = project.states.filter((st: AutomataState) => st.id !== state.id)
   })),
 
+  insertGroup: (createData) => {
+    // Check that we are inserting into same project type
+    if (createData.projectType !== get().project.projectType) {
+      return { type: InsertGroupResponseType.FAIL, body: `Error: you cannot insert elements from a ${createData.projectType} project into a ${get().project.projectType} project.` }
+    }
+    // Check that for transitions being inserted, to and from states are also inserted
+    let missingState = false
+    createData.transitions.forEach((transition) => {
+      let toFound = false
+      let fromFound = false
+      createData.states.forEach((state) => {
+        if (state.id === transition.from) {
+          fromFound = true
+        }
+        if (state.id === transition.to) {
+          toFound = true
+        }
+      })
+      // If a transition has to or from missing, flag to return error
+      if (!(toFound && fromFound)) {
+        missingState = true
+      }
+    })
+    if (missingState) {
+      return { type: InsertGroupResponseType.FAIL, body: 'Sorry, there was an error.' }
+    }
+    set(produce(({ project }: { project: StoredProject }) => {
+      let isInitialStateUpdated = false
+      const isNewProject = createData.projectSource !== project._id
+      const newTransitions = structuredClone(createData.transitions)
+      newTransitions.forEach(transition => {
+        transition.from = null
+        transition.to = null
+      })
+      createData.states.forEach((state, i) => {
+        // TODO: ensure position isn't out of window
+        // Probably will have to take adjusting position out of this function
+        state.x += PASTE_POSITION_OFFSET
+        state.y += PASTE_POSITION_OFFSET
+        const newId = i + 1 + Math.max(-1, ...get().project.states.map(s => s.id))
+        // Update transitions to new state id
+        createData.transitions.forEach((transition, i) => {
+          if (transition.from === state.id && newTransitions[i].from === null) {
+            newTransitions[i].from = newId
+          }
+          if (transition.to === state.id && newTransitions[i].to === null) {
+            newTransitions[i].to = newId
+          }
+        })
+        // Update initial state id if applicable
+        if (createData.initialStateId === state.id && !isInitialStateUpdated) {
+          createData.initialStateId = newId
+          isInitialStateUpdated = true
+        }
+        state.id = newId
+        // createState
+        project.states.push({ ...state })
+      })
+      createData.transitions = newTransitions
+      createData.comments.forEach((comment, i) => {
+        // TODO: ensure position isn't out of window
+        comment.x += PASTE_POSITION_OFFSET
+        comment.y += PASTE_POSITION_OFFSET
+        const newId = i + 1 + Math.max(-1, ...get().project.comments.map((c: ProjectComment) => c.id))
```

Comment:
Move this out into a function like
```ts
const nextIDFor = (ids: {id?: number}[]): number => i + 1 + Math.max(-1, ...ids.map(s => s.id))
```
And then use in the other places that creating a new ID is needed. 

Also not too sure but looks like that during these loops where a newID is needed the only difference is `i +`? So might be better to cache the highest found and then just increment/add `i` inside the loop